rust/kernel/spi.rs

```diff
@@ -0,0 +1,279 @@
+    fn shutdown(mut _spi_dev: SpiDevice) {}
+}
+
+/// Populate the TO_USE field in the [`SpiMethods`] implementer
+///
+/// ```rust
+/// impl SpiMethods for MySpiMethods {
+///     /// Let's say you only want a probe and remove method, no shutdown
+///     declare_spi_methods!(probe, remove);
+///
+///     /// Define your probe and remove methods. If you don't, default implementations
+///     /// will be used instead. These default implementations do NOT correspond to the
+///     /// kernel's default implementations! If you wish to use the Kernel's default
+///     /// spi functions implementations, do not declare them using the `declare_spi_methods`
+///     /// macro. For example, here our Driver will use the Kernel's shutdown method.
+///     fn probe(spi_dev: SpiDevice) -> Result {
+///         // ...
+///
+///         Ok(())
+///     }
+///
+///     fn remove(spi_dev: SpiDevice) -> Result {
+///         // ...
+///
+///         Ok(())
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! declare_spi_methods {
+    () => {
+        const TO_USE: $crate::spi::ToUse = $crate::spi::USE_NONE;
+    };
+    ($($method:ident),+) => {
+        const TO_USE: $crate::spi::ToUse = $crate::spi::ToUse {
+            $($method: true),+,
+            ..$crate::spi::USE_NONE
+        };
+    };
+}
+
+impl DriverRegistration {
+    fn new(this_module: &'static crate::ThisModule, name: &'static CStr) -> Self {
+        DriverRegistration {
+            this_module,
+            name,
+            registered: false,
+            spi_driver: bindings::spi_driver::default(),
+        }
+    }
+
+    /// Create a new `DriverRegistration` and register it. This is equivalent to creating
+    /// a static `spi_driver` and then calling `spi_driver_register` on it in C.
+    ///
+    /// ```
+    /// // The function returns a result, which you need to handle. Use the `?` operator
+    /// // to propagate it
+    /// let spi_driver =
+    ///     spi::DriverRegistration::new_pinned::<MySpiMethods>(&THIS_MODULE, cstr!("my_driver_name"))?;
+    /// ```
+    pub fn new_pinned<T: SpiMethods>(
+        this_module: &'static crate::ThisModule,
+        name: &'static CStr,
+    ) -> Result<Pin<Box<Self>>> {
+        let mut registration = Pin::from(Box::try_new(Self::new(this_module, name))?);
+
+        registration.as_mut().register::<T>()?;
+
+        Ok(registration)
+    }
+
+    unsafe extern "C" fn probe_wrapper<T: SpiMethods>(
+        spi_dev: *mut bindings::spi_device,
+    ) -> c_types::c_int {
+        // SAFETY: The spi_dev pointer is provided by the kernel and is sure to be valid
+        match T::probe(SpiDevice::from_ptr(spi_dev)) {
+            Ok(_) => 0,
+            Err(e) => e.to_kernel_errno(),
+        }
+    }
+
+    unsafe extern "C" fn remove_wrapper<T: SpiMethods>(
+        spi_dev: *mut bindings::spi_device,
+    ) -> c_types::c_int {
+        // SAFETY: The spi_dev pointer is provided by the kernel and is sure to be valid
```

Comment:
Use code quotes also in comments.

```suggestion
        // SAFETY: The `spi_dev` pointer is provided by the kernel and is sure to be valid.
```

Reply:
Fixed by 92a19c2ef2d0d0780c34327b273a21db627af711